### ISSUE TITLE
Inconsistency with 'sfp' SystemBlueprint object and o.r.core.sfp

### ISSUE THREAD

**Issue by albeanth** [terrapower/armi#876]
## The problem/observation
- in marching through `armi/reactor/tests/test_reactors.py::loadTestReactor`, there are two designs stored in `bp.systemDesigns`, `Core` and `sfp`. https://github.com/terrapower/armi/blob/55766354067d8b1cffed02a48d36cd5d59c10718/armi/reactor/reactors.py#L145-L147
- The `core` object is built just before this for loop here, https://github.com/terrapower/armi/blob/55766354067d8b1cffed02a48d36cd5d59c10718/armi/reactor/reactors.py#L143-L144. For `loadTestReactor` it gets 73 assemblies in core and 8 blueprints assemblies. 
- The `sfp` object is then built in `structure.construct`. It contains the 4 sfp assemblies defined in [the input](https://github.com/terrapower/armi/blob/main/armi/tests/refSmallSfpGrid.yaml) and 8 blueprints assemblies.

After both objects have been constructed, when `r` is returned the `sfp` object seems to be missing. 
- `r.sfp` is not a thing...
- `r.core.sfp` is an "AssemblyList object: Spent Fuel Pool", but contains no assemblies. 

Then in `loadTestReactor`, we put stuff into r.core.sfp https://github.com/terrapower/armi/blob/55766354067d8b1cffed02a48d36cd5d59c10718/armi/reactor/tests/test_reactors.py#L188-L191
- So, now we have these 10 assemblies in `r.core.sfp`, but the original 4 assemblies from the input are no where to be found (at least I can't find them). 

## Impacts
For the most part, I don't think this a huge deal (otherwise we would have found it by now). But, it does raise an issue if a user assigns assemblies in the SFP via the inputs or wants to see what assemblies are in the sfp on a database load.
- Since there isn't a registered `r.sfp` object and `r.core.sfp` doesn't get written to the database (at least in my tests it doesn't), then there is no way to check what is actually in the sfp on a database load (either put in via a shuffle or initialized there via an input).
  - I imagine this is certainly something we'll want to support. Knowing what is in the SFP feels important. Especially when we're shuffling assemblies out of the core. 
- There is a disconnect between where ARMI stores SFP assemblies from inputs (in Core: sfp object) vs where it stashes them as a result of shuffling (in r.core.sfp). 

- This also wreaks havoc a bit on the global assem counter as it has been ticked to account for SFP assemblies, but they aren't anywhere to be found. https://github.com/terrapower/armi/blob/55766354067d8b1cffed02a48d36cd5d59c10718/armi/bookkeeping/db/tests/test_database3.py#L331-L351

**Comment by mgjarrett** [comment]
I think @keckler uses the SFP more than anyone else, so he might have some insight.

**Comment by john-science** [comment]
And this has nothing to do with our recent removal of the blueprints from `db.load()`?

**Comment by jakehader** [comment]
@john-science / @albeanth - I think that this might be an issue with the database loading based on this comment that has been in the code for some time

https://github.com/terrapower/armi/blob/main/armi/bookkeeping/db/database3.py#L1224-L1231

**Comment by onufer** [comment]
the issue seems to be the assemblies in  SFP are created but just dropped on the ground and not actually stored in sfp.

When test_reactor is created it should have 4 assemblies in SFP according to the blueprints grid, and they are created, but they are never stored on the object
![image](https://user-images.githubusercontent.com/50242044/190243446-020ea63e-ff72-4dba-8707-a97fbccc2519.png)


AssertionError: assert 0 == 4

**Comment by mgjarrett** [comment]
The `r.core.sfp` is a `armi.reactor.assemblyLists.AssemblyList`. 

The "sfp" defined in the blueprints is constructed as a `armi.reactor.reactors.Core` object. See example output from `test_database3::TestDatabase3::test_load_updateGlobalAssemNum` below:
`=========== Adding Assemblies to <Core: sfp id:2962777429048> ===========`

I would be surprised if these were supposed to be one and the same when initially implemented because they are two distinct types. Not sure what the intention was for the blueprints `sfp`, but if it was for those assemblies to ultimately end up in `r.core.sfp` then it's not even half-baked.

If the `Core` object named `sfp` hangs around long enough, we could iterate through `sfp.getAssemblies()` and `add` those assemblies to the `r.core.sfp`. Not sure at what point the `sfp` gets created and how long it persists before it's tossed, but that might be possible without any significant code changes.

**Comment by albeanth** [comment]
> I think that this might be an issue with the database loading based on this comment that has been in the code for some time

It's definitely a piece of the puzzle. I believe that `if` statement is there because `Core: sfp` is a Core object and doing processLoading on `Core: sfp` doesn't make sense (processLoading is a `Core: core` specific operation). If we treated the SFP as its own class (maybe a subclass of Core?) and stored it on the Reactor we could likely get rid of that TODO and if statement (it would never get there since comp would be an Sfp and not a Core). 

We could store the sfp on the Reactor like this:
```python
class Reactor(composites.Composite):
    ...
    def __init__(self, name, blueprints):
        ...
        self.core = None
        self.blueprints = blueprints
        self.sfp = None
```

------
> the issue seems to be the assemblies in SFP are created but just dropped on the ground and not actually stored in sfp.

It's definitely another piece of the puzzle. The downside though is that even if we get the assemblies in the SFP stored correctly on `r.core.sfp` they won't get written to the database. It might be unnecessary, but should we have a separate SFP class? One that's analogous to the Core class we have now? Maybe a subclass of Core. 
 
------

> I would be surprised if these were supposed to be one and the same when initially implemented because they are two distinct types. 

I'd argue that they probably shouldn't be the same type. They share characteristics, sure, but I'm not sure if they need to/should  be the same types. They need to have overlap in case someone want so to do analysis on only an SFP, but I'm not so sure they should be the exact same type. I could certainly be wrong though.

> Not sure what the intention was for the blueprints sfp

This is a byproduct of the `Core: sfp` being a "Core" object. Since it is, it goes through the same construction process as a "regular" Core and as such gets blueprints via `blueprints/__init__.py/Blueprints/_prepConstruction`. These are then used to "fill" the sfp, just like a regular Core is. It might be more work than it's worth, but I think it would make sense to build `r.blueprints` first and then build the Core and/or SFP (both as their own classes).

> we could iterate through sfp.getAssemblies() and add those assemblies to the r.core.sfp

This is something we could do. It would at least get all (input-based and programmatic) of the sfp assemblies in one spot. But it would not resolve the db issue of the sfp assemblies missing altogether with. Since `r.core.sfp` is not considered a "child" of `r.core`, those assemblies never get written to the database.

**Comment by albeanth** [comment]
> And this has nothing to do with our recent removal of the blueprints from `db.load()`?

I don't believe so. In #872 we ended up keeping optional blueprints param in `db.load`. To prevent users from loading the wrong blueprints into a database (i.e., blueprints that don't match what's on the db), we removed passing in the optional parameter in `loadState`.

This issue is more of a problem with where SFP assemblies are (or are not) stored and that they are never written to the db.

**Comment by albeanth** [comment]
Was poking around as part of a PR review and came across something relevant to this issue. 

It seems as though the framework setting, `trackAssems`, was intended to enable the saving of discharged assemblies in the sfp.
https://github.com/terrapower/armi/blob/6684ae415898f8986ed5e680bfc4b0e417559d30/armi/settings/fwSettings/globalSettings.py#L672-L680

Unfortunately, since they get stashed in `r.core.sfp` they won't actually get written out to the database. So the increased memory requirements that come with enabling this setting may not actually be worth much.

**Comment by keckler** [comment]
> So the increased memory requirements that come with enabling this setting may not actually be worth much.

Well, I would disagree with this statement. Just because they aren't written to the database doesn't mean that the SFP assemblies can't still be used during the run. So in a non-restart run, they are definitely still valuable.

**Comment by albeanth** [comment]
> > So the increased memory requirements that come with enabling this setting may not actually be worth much.
> 
> Well, I would disagree with this statement. Just because they aren't written to the database doesn't mean that the SFP assemblies can't still be used during the run. In a non-restart run, they can definitely still be valuable.

Good point!

**Comment by keckler** [comment]
I agree with @albeanth that the assemblies in the "SFP" (whatever form that might take) need to be written to the DB. I was not aware that `r.core.sfp` was not put into the DB, but looking into it now, I agree that it is empty when I reload a reactor from a DB that previously had discharged assemblies.

I think it is very important that `r.core.sfp` should be properly repopulated when loading from a DB. As @albeanth points out, if a user turns on `trackAssems`, presumably those discharged assemblies are important to them.

In order to do this, I think we will have to modify the [`_createLayout`](https://github.com/terrapower/armi/blob/main/armi/bookkeeping/db/database3.py#L2060) and [`_readLayout`](https://github.com/terrapower/armi/blob/main/armi/bookkeeping/db/database3.py#L2117) methods of the `Layout` class. Possibly others will need to be modified as well. Might get a little hairy with making sure the reactor model is restored properly with the addition.

To me, the best way would be to keep the discharged assemblies stored as just an `AssemblyList`, and to not have a separate `Core` object. No real reason for that, except that if it _were_ a `Core` object, then the user would have to care about which position the assembly is put into when it is discharged from the real core, which is just probably an unnecessary complication.

This is a real problem, so thanks for the discovery @albeanth !

**Comment by ntouran** [comment]
SFP was around before the reactor could have child _systems_ and is kind of shoe-horned in for transitional reasons. It's supposed to end up being just another _system_ alongside the core as a child of the `reactor`. And `r.sfp` might be useful as a convenient shortcut for that child, kind of like how `r.core` is a convenient shortcut for the core system.

**Comment by albeanth** [comment]
Thank you for the insight, @ntouran !

**Comment by albeanth** [comment]
> To me, the best way would be to keep the discharged assemblies stored as just an AssemblyList, and to not have a separate Core object. No real reason for that, except that if it were a Core object, then the user would have to care about which position the assembly is put into when it is discharged from the real core, which is just probably an unnecessary complication.

Instead of making it an instance of Core, I think it makes sense to have the SFP be its own class that lives on `r.sfp`.  I think there's some logic in maybe subclassing Core for an Sfp class (e.g., I could imagine one wanted to do criticality analyses on a sfp), but would need some thought. 

Not sure I completely understand the assembly position complication you bring up. Can you clarify @keckler ?

**Comment by keckler** [comment]
It's not a big deal at all. I'm just thinking... Right now when I remove an assembly from the core and put it in `r.core.sfp`, its just in some big list.

If instead I had to put the discharged assembly into gridded system (analogous to a `Core`, but maybe different), then the assembly would have to go in a specific position. For instance, position 1A in a cartesian grid or something like that. 

Obviously we could make it so that the assembly just gets placed in the next available position, so my concern is effectively trivial if we were to implement a new class specific to SFPs.

I very much like the idea of having a specific SFP class, though the design of such a class would need some thought. As you've seen, hardly anyone uses the SFP in ARMI right now, so I imagine the class could be pretty bare-bones. It might just be a matter of having a grid that gets stored in the DB, with no other frills for now.

**Comment by albeanth** [comment]
Ah, that makes total sense. We're on the same page now. A solid observation!

**Comment by john-science** [comment]
@mgjarrett Any updates on this ticket, after your recent SFP PR?

**Comment by mgjarrett** [comment]
> @mgjarrett Any updates on this ticket, after your recent SFP PR?

Somehow I completely missed this comment until _just now_... yes, I believe that this ticket should have been closed when #1349 was merged.
